core/src/block.rs

```diff
@@ -253,17 +250,146 @@ mod valid {
     pub struct ValidBlock(pub(super) SignedBlock);
 
     impl ValidBlock {
+        fn verify_leader_signature(
+            block: &SignedBlock,
+            topology: &Topology,
+        ) -> Result<(), SignatureVerificationError> {
+            let leader_index = topology.leader_index();
+            let mut block_signatures = block.signatures();
+
+            let leader_signature = match block_signatures.next() {
+                Some(BlockSignature(signatory, signature))
+                    if usize::try_from(*signatory)
+                        .map_err(|_err| SignatureVerificationError::LeaderMissing)?
+                        == leader_index =>
+                {
+                    let mut additional_leader_signatures =
+                        topology.filter_signatures_by_roles(&[Role::Leader], block_signatures);
+
+                    if additional_leader_signatures.next().is_some() {
+                        return Err(SignatureVerificationError::DuplicateSignatures {
+                            signatory: leader_index,
+                        });
+                    }
```

Comment:
malicious peer can send all sorts of invalid blocks. What you're saying is true, but is no less dangerous than sending other types of invalid blocks. We just reject a block sent from this peer, there is no harm in that IMO

Reply:
> I think what @Erigara is saying is that the block validation will globally and repeatedly fail unless it has a mechanism to identify and ignore (or expel from the validator set) the peer that spoof index zero with its signature.

spoofing 0 index is just one way in which malicious peer can affect consensus. Worst case, malicious peer can withhold sending any messages to other peers. It will only fail consensus repeatedly until the malicious peer becomes observing peer. 0 index spoofing is no special case and I don't think we should be more forgiving in this case

Reply:
Ok, at least this is not a problem specific to this PR. It used to trust public keys instead of indices